tests/main/views/test_suppliers.py

```diff
@@ -1932,3 +1932,298 @@ def test_agree_variation_various(
         assert audit_events[1].data["update"] == {
             "agreedUserId": 2,
         }
+
```

Comment:
So I'm not going to suggest we rewrite these tests using model creation but I do want to take a bit of time to offer a critique of why I think they have some drawbacks.

First off, disclaimer, and I know they were inspired by the code that tests the user export, so it was a totally logical choice to use this pattern here and one which I think was, given the info at hand a good choice. 
I'm seriously all for following existing patterns and trusting previous bits of code. So in many ways using this pattern was the right thing to do. _That said_ I don't think it's a great pattern and here's why:

The specific problem I have with these tests is how they set up their test data.

In many cases they use views that already exist to set up test data. This might seem like a logical thing to do and really it sort of is. That's what the views are for after all. But it comes with quite a big drawback. They test _too much_.

These tests now will fail if someone changes something in the views on which they rely. Therefore they test not only the `supplier_export` but also all the associated views they hit. If they can fail when I change something then they are testing that thing.

So why does that matter? If I change something in the `draft-services` end point and tests in `test_suppliers` start failing then thats extra noise I really don't need and it detracts from the overall usefulness of the test suite. I have to find out why they're failing, and this happened with the debugging @katstevens had to do to find out what was happening to the framework object.

Basically if a unit test tests _everything_ then they don't test any one thing very effectively. 

Really useful unit tests obviously test only one unit of code. And one part of the codebase.
For these tests specifically that part of the codebase should be the relationship/ contract between what's in the database and what comes out of the api. It would only be testing the code between the database and the end of the view function (e the view function).

So if we only want to test the relationship between what's in the DB and what comes out of the api (so we know that the `export_suppliers` view has been written correctly) then how should we do it?

Have some input data (some test data that goes directly into the DB). Put it through our process (call the view) and compare the output to a set of expected values.

That way we know that whatever is causing the failures is in the code we wrote.

It's really tempting to test as much code as possible with unit tests but it makes them less useful overall if they fail every time a piece of code changes.
If you change a piece of code in the `supplier_export` view and 30 tests break then that is not useful. 
If you change a piece of code in the `supplier_export` view and 1 test called `test_correct_supplier_export_headings` breaks then that is super useful. 

Just my 2 cents. And as I say don't change them now!

Reply:
I have to agree that I don't like using API methods in this way to set up test data. While it's a nice simulation of how the data gets into that state, we... don't care about that for these tests? We shouldn't make a habit of this pattern IMHO.

Reply:
> Basically if a unit test tests everything then they don't test any one thing very effectively.
> Really useful unit tests obviously test only one unit of code.

Though I agree that it's best not to use api endpoints to do your setup (mostly because of performance), you've made a broader philosophical statement so you're going to get a broader philosophical response...

I strongly disagree with dogmatic beliefs about the "technical" difference between unit tests and functional-integration-whatever-you-want-to-call-it-tests and don't think we should be propagating them. Nothing is purely a unit test. Nothing is purely a functional test. It's a spectrum. And also what is most effective in any particular situation varies massively. The important thing is whether it asserts things that you care about. The highly "technical" differences are largely made up by people who want to sell books. The reality is that if all you have is "pure" unit tests, you will *never* write enough of them to be able to properly assert the correct high level behaviour of your system and at that point they will take ages to run. They will also be close to useless come refactoring time. The same amount of code can usually be exercised in a much smaller & less fragile amount of test code with a *higher level* test. Testing is a place for pragma.

Insert "two passing unit tests, zero integration tests" image macro of kitchen drawers here.